### PR TITLE
Replace libplatform and libtapasco with Rust based modern version

### PR DESCRIPTION
**Pull request by jahofmann**
This pull request introduces a new Rust based library that is much easier to use and has better error handling compared to the traditional version.

In addition this runtime is easier to extend with features in future releases.

Overall performance is better in most cases and especially when using multiple threads:
[performance_rust.pdf](https://github.com/esa-tu-darmstadt/tapasco/files/4723507/performance_rust.pdf)

Downsides:
 - The C API has changed. Hopefully, most people use the C++ API, so this is not an issue.
 - Initial compile time can be higher on slow machines (Around 45s on my development VM). Subsequent compiles are much faster and tools like https://github.com/mozilla/sccache provide additional compile time reductions. This is not a problem for the end user as they are expected to use the OS packets which contain already compiled libraries.

### CODE REVIEW

File: runtime/examples/C/arrayinit/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,17 @@
+cmake_minimum_required(VERSION 3.5.1 FATAL_ERROR)
```
Comment (jojowi) [inline runtime/examples/C/arrayinit/CMakeLists.txt:1]: Our new copyright header is missing

File: runtime/examples/C/arraysum/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,16 @@
+cmake_minimum_required(VERSION 3.5.1 FATAL_ERROR)
```
Comment (jojowi) [inline runtime/examples/C/arraysum/CMakeLists.txt:1]: Copyright Header

File: runtime/examples/C/arrayupdate/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,16 @@
+cmake_minimum_required(VERSION 3.5.1 FATAL_ERROR)
```
Comment (jojowi) [inline runtime/examples/C/arrayupdate/CMakeLists.txt:1]: Copyright header

File: runtime/examples/C/cffitest/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,10 @@
+cmake_minimum_required(VERSION 3.0)
```
Comment (jojowi) [inline runtime/examples/C/cffitest/CMakeLists.txt:1]: Copyright

File: runtime/examples/C/cffitest/ctest.c [hunk 5]
```diff
@@ -0,0 +1,106 @@
+#include <stdio.h>
```
Comment (jojowi) [inline runtime/examples/C/cffitest/ctest.c:1]: Copyright

File: runtime/examples/Rust/libtapasco_tests/src/main.rs [hunk 6]
```diff
@@ -0,0 +1,687 @@
+extern crate crossbeam;
```
Comment (jojowi) [inline runtime/examples/Rust/libtapasco_tests/src/main.rs:1]: Copyright

File: runtime/libtapasco/CMakeLists.txt [hunk 7]
```diff
@@ -0,0 +1,59 @@
+cmake_minimum_required(VERSION 3.0)
```
Comment (jojowi) [inline runtime/libtapasco/CMakeLists.txt:1]: Copyright

File: runtime/libtapasco/TapascoInner.cmake [hunk 8]
```diff
@@ -0,0 +1,31 @@
+set(CARGO_GREP grep -oP \"note: native-static-libs: \\K.+\")
```
Comment (jojowi) [inline runtime/libtapasco/TapascoInner.cmake:1]: Copyright

File: runtime/libtapasco/includes/libtapasco.h [hunk 9]
```diff
@@ -0,0 +1,12 @@
+#ifndef LIBTAPASCO_H_
```
Comment (jojowi) [inline runtime/libtapasco/includes/libtapasco.h:1]: Copyright?

File: runtime/libtapasco/src/allocator.rs [hunk 10]
```diff
@@ -0,0 +1,296 @@
...
+                }
+                self.merge_memory();
+                Ok(())
+            }
+            None => Err(Error::UnknownMemory { ptr: ptr }),
+        }
+    }
+}
+
+#[test]
```
Comment (jojowi) [inline runtime/libtapasco/src/allocator.rs:188]: Is there a reason theses tests are here (and in examples/Rust/libtapasco_tests)?

File: runtime/libtapasco/src/allocator.rs [hunk 11]
```diff
@@ -0,0 +1,296 @@
...
+}
+
+#[derive(Debug, Getters, Copy, Clone)]
+struct MemoryFree {
+    base: DeviceAddress,
+    size: DeviceSize,
+}
+
+#[derive(Debug, Getters)]
+pub struct GenericAllocator {
```
Comment (jojowi) [inline runtime/libtapasco/src/allocator.rs:59]: Some comments would be nice. Especially for the difference between GenericAllocator and DriverAllocator

File: runtime/libtapasco/src/dma.rs [hunk 12]
```diff
@@ -0,0 +1,179 @@
...
+}
+type Result<T, E = Error> = std::result::Result<T, E>;
+
+pub trait DMAControl: Debug {
+    fn copy_to(&self, data: &[u8], ptr: DeviceAddress) -> Result<()>;
+    fn copy_from(&self, ptr: DeviceAddress, data: &mut [u8]) -> Result<()>;
+}
+
+#[derive(Debug, Getters)]
+pub struct DriverDMA {
```
Comment (jojowi) [inline runtime/libtapasco/src/dma.rs:65]: Again, some comments regarding the difference between DriverDMA and DirectDMA would be good

File: runtime/libtapasco/src/tapasco.hpp [hunk 13]
```diff
@@ -0,0 +1,579 @@
...
+  }
+
+  /**
+   * Allocates a chunk of len bytes on the device.
+   * @param len size in bytes
+   * @param h output parameter for handle
+   * @param flags device memory allocation flags
+   * @return TAPASCO_SUCCESS if successful, an error code otherwise.
+   **/
+  int alloc(DeviceAddress &h, size_t const len) {
```
Comment (jojowi) [inline runtime/libtapasco/src/tapasco.hpp:441]: Here (alloc, free, copy_to, copy_from) the C++-API changed. All methods had an additional flag-argument, free() also had an argument for the size of the memory which should be freed.
I see these arguments are no longer necessary, but maybe we should also keep the methods with the old signatures and just ignore the additional arguments? Then we wouldn't brake old code

File: toolflow/vivado/platform/pcie/pcie_base.tcl [hunk 14]
```diff
@@ -281,28 +281,10 @@
...
-      } else {
-        puts "Using PCIe IP for x16..."
-        set bridge [tapasco::ip::create_pciebridgetolite_x16 "PCIeBridgeToLite"]
-      }
-        connect_bd_intf_net [get_bd_intf_pins -regexp $pcie/M_AXI(_B)?] \
-      [get_bd_intf_pins -of_objects $bridge -filter "VLNV == [tapasco::ip::get_vlnv aximm_intf] && MODE == Slave"]
-
-      connect_bd_net [get_bd_pins axi_pcie3_0/axi_aclk] [get_bd_pins -of_objects $bridge -filter "TYPE == clk"]
-      connect_bd_net [get_bd_pins axi_pcie3_0/axi_aresetn] [get_bd_pins -of_objects $bridge -filter "TYPE == rst"]
+    tapasco::ip::connect_sc_default_clocks $out_ic "host"
```
Comment (jojowi) [inline toolflow/vivado/platform/pcie/pcie_base.tcl:284]: Why are the BridgeToLite-IPs no longer required? How is this related to the new runtime?

File: runtime/libtapasco/src/allocator.rs (hunk 11 above)
Comment (jahofmann) [reply runtime/libtapasco/src/allocator.rs:59]: Yes! Comments are missing almost everywhere right now...

File: runtime/libtapasco/src/allocator.rs (hunk 10 above)
Comment (jahofmann) [reply runtime/libtapasco/src/allocator.rs:188]: The tests are used wherever I felt that they are needed and can be implemented easily right now.

I would love to have more tests, but currently the tests are limited to parts of the code that don't interact with the driver. As soon as we have a dummy version of the driver I'd like to change that.

File: runtime/libtapasco/src/tapasco.hpp (hunk 13 above)
Comment (jahofmann) [reply runtime/libtapasco/src/tapasco.hpp:441]: Yes!

File: toolflow/vivado/platform/pcie/pcie_base.tcl (hunk 14 above)
Comment (jahofmann) [reply toolflow/vivado/platform/pcie/pcie_base.tcl:284]: Because the Rust functions I used are much more efficient and use e.g. SSE transfers whenever available they will break the forced conversion to 32 Bit/64Bit which we used previously.

File: runtime/libtapasco/src/allocator.rs (hunk 10 above)
Comment (jojowi) [reply runtime/libtapasco/src/allocator.rs:188]: I forgot a "not" in my question. I wanted to say "and **not** in examples/Rust/libtapasco_tests?".
Is there a reason these tests are directly in the source file and not with the rest of the tests in libtapasco_tests?

Comment (jahofmann) [reply runtime/libtapasco/src/allocator.rs:188]: Ohh, those are different kind of tests.

The `libtapasco_tests` is used to do some initial testing on the device, but doesn't really test the code itself.

These tests here marked `#[test]` are from the unit testing framework provided by Rust and are run using `cargo test`. They are supposed to be located right next to the code they affect and can be used wherever the programmer feels that such tests are necessary. There are also corresponding `benchmark` versions for micro benchmarks which would be really useful (as soon as we have a dummy driver).

Comment (jojowi) [reply runtime/libtapasco/src/allocator.rs:188]: Ok

File: runtime/libtapasco/src/tlkm.rs [hunk 15]
```diff
@@ -2,7 +2,7 @@
  * Copyright (c) 2014-2020 Embedded Systems and Applications, TU Darmstadt.
  *
  * This file is part of TaPaSCo
- * (see https://github.com/esa-tu-darmstadt/tapasco).
+ * (see https:///github.com/esa-tu-darmstadt/tapasco).
```
Comment (m-ober) [inline runtime/libtapasco/src/tlkm.rs:5]: Why three slashes? (also in the last line)

Comment (jahofmann) [reply runtime/libtapasco/src/tlkm.rs:5]: Good question

**Comment by jahofmann** [comment]
These last few commits add eventfd based interrupt handling for the design interrupts, which greatly simplifies the driver and runtime. Performance seems similar, but I have to do further testing to confirm this.

**Comment by jahofmann** [comment]
I did some more testing and added a few low latency improvements with huge success. As shown in the graph the latency over the full range dropped by over 50% and the number of Jobs increased significantly as well.

[performance_eventfd.pdf](https://github.com/esa-tu-darmstadt/tapasco/files/4798005/performance_eventfd.pdf)

**Comment by kmeinhar** [comment]
If I am not mistaken the "Prerequisites for the Runtime" section in the top  level README.me should now also include the following packages: `rustc`, `cargo` and `protobuf`

**Comment by jahofmann** [comment]
Totally! The documentation is simply not updated at all. Will do that after the next deadlines...

To get started you should use https://rustup.rs to install Rust and the rest.

File: runtime/examples/Rust/libtapasco_tests/src/main.rs [hunk 16]
```diff
@@ -0,0 +1,687 @@
...
+    Ok(())
+}
+
+fn run_arrayinit(_: &ArgMatches) -> Result<()> {
+    let tlkm = TLKM::new().context(TLKMInit {})?;
+    let devices = tlkm.device_enum(&HashMap::new()).context(TLKMInit)?;
+    for mut x in devices {
+        x.change_access(tapasco::tlkm::tlkm_access::TlkmAccessExclusive)
+            .context(DeviceInit {})?;
+        let mut pe = x.acquire_pe(11).context(DeviceInit)?;
```
Comment (kmeinhar) [inline runtime/examples/Rust/libtapasco_tests/src/main.rs:120]: Using magic numbers can be confusing.
Create a constant integer with a value at the top:
```
       const ARRAY_PE_ID: u32 = 1745;
```
```suggestion
        let mut pe = x.acquire_pe(ARRAY_PE_ID).context(DeviceInit)?;
```

Comment (jahofmann) [reply runtime/examples/Rust/libtapasco_tests/src/main.rs:120]: Yes, I am not a fan of these numbers as well. For the release I will replace them completely. Then you can request the ID from the API, something like: `device.getPEId("ARRAY_INIT")`. This also makes moving between bitstreams generated by different people (maybe using different IDs) easier.

**Comment by jahofmann** [comment]
The latest commits add "get ID by VLNV" and dynamic interrupts.

This makes this pull request feature complete (I hope).

I will spend a day to finalize documentation and then this should be good to go.

File: toolflow/vivado/common/debug.patch [hunk 17]
```diff
@@ -0,0 +1,62 @@
+diff --git a/toolflow/vivado/common/ip.tcl b/toolflow/vivado/common/ip.tcl
```
Comment (jojowi) [inline toolflow/vivado/common/debug.patch:1]: Is there a reason this is here?

**Review by jojowi — APPROVED** [review]
No further comments from my side
